recipe/0001-drop-pyarrow-version-numbers.patch

```diff
@@ -4,70 +4,173 @@ Date: Sat, 30 Apr 2022 10:16:22 -0400
 Subject: [PATCH] drop pyarrow version numbers
 
 ---
- pyproject.toml |  2 +-
- setup.cfg      |  2 +-
- setup.py       | 16 ++++++++--------
- 3 files changed, 10 insertions(+), 10 deletions(-)
+ pyproject.toml                                |  2 +-
+ setup.cfg                                     |  2 +-
+ setup.py                                      | 52 +++++++++++--------
+ .../cpp/ArrowIterator/BinaryConverter.cpp     |  4 ++
+ .../cpp/ArrowIterator/StringConverter.cpp     |  4 ++
+ 5 files changed, 41 insertions(+), 23 deletions(-)
 
 diff --git a/pyproject.toml b/pyproject.toml
-index 87ddbe1..6d4266e 100644
+index 2bdafcb5..6d4266ef 100644
 --- a/pyproject.toml
 +++ b/pyproject.toml
 @@ -6,7 +6,7 @@ requires = [
      "wheel",
      "cython",
      # Must be kept in sync with the `setup_requirements` in `setup.cfg`
--    "pyarrow>=8.0.0,<8.1.0",
+-    "pyarrow>=10.0.1,<10.1.0",
 +    "pyarrow",
  ]
- 
+
  [tool.cibuildwheel]
 diff --git a/setup.cfg b/setup.cfg
-index fc4d873..e6b0c0d 100644
+index 1b98b685..0fb7c12f 100644
 --- a/setup.cfg
 +++ b/setup.cfg
-@@ -91,6 +91,6 @@ development =
+@@ -92,6 +92,6 @@ development =
      pytzdata
  pandas =
      pandas>=1.0.0,<1.6.0
--    pyarrow>=8.0.0,<8.1.0
+-    pyarrow>=10.0.1,<10.1.0
 +    pyarrow
  secure-local-storage =
      keyring!=16.1.0,<24.0.0
 diff --git a/setup.py b/setup.py
-index d4ccdb7..53472f9 100644
+index 09336043..4861a9cd 100644
 --- a/setup.py
 +++ b/setup.py
-@@ -71,21 +71,21 @@ if _ABLE_TO_COMPILE_EXTENSIONS:
-         # this list should be carefully examined when pyarrow lib is
+@@ -48,8 +48,9 @@ try:
+     from Cython.Distutils import build_ext
+
+     _ABLE_TO_COMPILE_EXTENSIONS = True
+-except ImportError:
++except ImportError as e:
+     warnings.warn("Cannot compile native C code, because of a missing build dependency")
++    raise
+     _ABLE_TO_COMPILE_EXTENSIONS = False
+
+ if _ABLE_TO_COMPILE_EXTENSIONS:
+@@ -72,16 +73,16 @@ if _ABLE_TO_COMPILE_EXTENSIONS:
          # upgraded
          arrow_libs_to_copy = {
--            "linux": ["libarrow.so.800", "libarrow_python.so.800", "libparquet.so.800"],
-+            "linux": ["libarrow.so", "libarrow_python.so", "libparquet.so"],
+             "linux": [
+-                "libarrow.so.1000",
+-                "libarrow_dataset.so.1000",
+-                "libarrow_python.so.1000",
+-                "libparquet.so.1000",
++                "libarrow.so",
++                "libarrow_dataset.so",
++                "libarrow_python.so",
++                "libparquet.so",
+             ],
```

Comment:
There are several things wrong with this:
* no copying of libs should happen (might be necessary for wheels, but nono for conda-forge)
* you're copying the symlinks (pointing to the versioned libs) without copying what the link points to.

Reply:
IIUC, we're setting SF_NO_COPY_ARROW_LIB=1 to disable the actual copying, this patch just allows for building against multiple versions of arrow. 

IMO, as a maintainer it would be easier if we only needed to support one version of arrow to match what the upstream library actually supports. This is done so we can support many versions of arrow.

Reply:
> IMO, as a maintainer it would be easier if we only needed to support one version of arrow to match what the upstream library actually supports. This is done so we can support many versions of arrow.

Supporting only one version of arrow is convenient but very user-hostile. If snowflake pinned to 9 and another package pinned to 10, there wouldn't be a way to co-install them. Hard pins should really be a last resort, and with the exception of the library restructuring with arrow 10, there shouldn't be much of an issue to support several arrow versions.

The alternative is either user pain or a real degradation in usability of the snowflake connector.

Reply:
Understood. Prior to this there were no actual library differences needed but now it seems we will need to maintain a patch to keep the backwards compatibility for arrow <10. Hopefully that won't be too long lived.